File: android-app/app/src/main/java/com/kvnetflix/mobile/viewmodel/PairingViewModel.kt
package com.kvnetflix.mobile.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.kvnetflix.mobile.data.api.ApiClient
import com.kvnetflix.mobile.data.repository.UserDataRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

sealed class PairingUiState {
    object Idle : PairingUiState()
    object Loading : PairingUiState()
    object Success : PairingUiState()
    data class Error(val message: String) : PairingUiState()
}

class PairingViewModel(private val userDataRepository: UserDataRepository) : ViewModel() {

    private val _uiState = MutableStateFlow<PairingUiState>(PairingUiState.Idle)
    val uiState: StateFlow<PairingUiState> = _uiState.asStateFlow()

    fun loginWithCode(code: String) {
        if (code.length < 6) return

        viewModelScope.launch {
            _uiState.value = PairingUiState.Loading
            try {
                val response = ApiClient.api.loginWithCode(mapOf("code" to code))
                ApiClient.authToken = response.token
                userDataRepository.saveAuthData(response.token, response.user)
                userDataRepository.syncWithRemote()
                _uiState.value = PairingUiState.Success
            } catch (e: retrofit2.HttpException) {
                val errorMsg = when (e.code()) {
                    400 -> "Invalid or expired code."
                    404 -> "Code not found."
                    else -> "Connection error (${e.code()})"
                }
                _uiState.value = PairingUiState.Error(errorMsg)
            } catch (e: Exception) {
                _uiState.value = PairingUiState.Error(e.message ?: "Failed to login")
            }
        }
    }

    fun resetState() {
        _uiState.value = PairingUiState.Idle
    }
}

File: ios-app/KvNetflix/ViewModels/PairingViewModel.swift
import Foundation
import Combine

public enum PairingState: Equatable {
    case idle
    case loading
    case success
    case error(String)
}

@MainActor
public final class PairingViewModel: ObservableObject {
    @Published public private(set) var state: PairingState = .idle
    @Published public var code: String = ""
    @Published public private(set) var generatedLinkCode: String?
    @Published public private(set) var isGeneratingCode: Bool = false

    private let api = ApiClient.shared
    private let userRepo = UserDataRepository.shared

    public init() {}

    public func loginWithCode(_ codeString: String? = nil) {
        let targetCode = (codeString ?? code).trimmingCharacters(in: .whitespacesAndNewlines)
        guard targetCode.count >= 6 else {
            state = .error("Please enter a valid 6-digit code")
            return
        }

        state = .loading

        Task {
            do {
                let response = try await api.loginWithCode(code: targetCode)
                userRepo.saveAuthData(token: response.token, profile: response.user)
                await userRepo.syncWithRemote()
                self.state = .success
            } catch let nsError as NSError {
                let msg: String
                switch nsError.code {
                case 400: msg = "Invalid or expired pairing code."
                case 404: msg = "Code not found."
                default: msg = nsError.localizedDescription
                }
                self.state = .error(msg)
            }
        }
    }

    public func generateLinkCode() {
        isGeneratingCode = true
        Task {
            do {
                let code = try await userRepo.generateLinkCode()
                self.generatedLinkCode = code
                self.isGeneratingCode = false
            } catch {
                self.isGeneratingCode = false
                self.state = .error("Failed to generate code: \(error.localizedDescription)")
            }
        }
    }

    public func resetState() {
        state = .idle
        code = ""
    }
}
